File: src/roar_net_api/operations/moves.py
# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0

from collections.abc import Iterable
from typing import Protocol, TypeVar

Solution = TypeVar("Solution", contravariant=True)
Move = TypeVar("Move", covariant=True)


class SupportsMoves(Protocol[Solution, Move]):
    def moves(self, solution: Solution) -> Iterable[Move]: ...

File: julia/src/operations/moves.jl
# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0
"""
    moves(neighbourhood::Neighbourhood, solution::Solution) -> iterable::Any

Return an iterable of all valid `Move`s in `neighbourhood` for `solution`.
The return type is declared `Any` because concrete implementations may use
either a `Vector{<:Move}` or a `Channel{<:Move}`.
"""
function moves(::Neighbourhood, ::Solution)::Any end
